src/services/pasteEdits.ts

```diff
@@ -0,0 +1,103 @@
+import {
+    addRange,
+} from "../compiler/core";
+import {
+    CancellationToken,
+    SourceFile,
+    Statement,
+    SymbolFlags,
+    TextRange,
+    UserPreferences,
+} from "../compiler/types";
+import {
+    getLineOfLocalPosition,
+} from "../compiler/utilities";
+import {
+    codefix,
+    fileShouldUseJavaScriptRequire,
+    forEachChild,
+    formatting,
+    getQuotePreference,
+    getTargetFileImportsAndAddExportInOldFile,
+    insertImports,
+    isIdentifier,
+    textChanges,
+} from "./_namespaces/ts";
+import {
+    getExistingLocals,
+    getUsageInfo,
+} from "./refactors/moveToFile";
+import {
+    CodeFixContextBase,
+    FileTextChanges,
+    LanguageServiceHost,
+    PasteEdits,
+} from "./types";
+
+/** @internal */
+export function pasteEditsProvider(
+    targetFile: SourceFile,
+    copies: { text: string; copyRange?: { file: SourceFile; range: TextRange; }; }[],
+    pastes: TextRange[],
+    host: LanguageServiceHost,
+    preferences: UserPreferences,
+    formatContext: formatting.FormatContext,
+    cancellationToken: CancellationToken,
+): PasteEdits {
+    const changes: FileTextChanges[] = textChanges.ChangeTracker.with({ host, formatContext, preferences }, changeTracker => pasteEdits(targetFile, copies, pastes, host, preferences, formatContext, cancellationToken, changeTracker));
+    return { edits: changes };
+}
+
+function pasteEdits(
+    targetFile: SourceFile,
+    copies: { text: string; copyRange?: { file: SourceFile; range: TextRange; }; }[],
+    pastes: TextRange[],
+    host: LanguageServiceHost,
+    preferences: UserPreferences,
+    formatContext: formatting.FormatContext,
+    cancellationToken: CancellationToken,
+    changes: textChanges.ChangeTracker,
+) {
+    const copy = copies[0];

```

Comment:
Have we decided to ignore multiple copy locations? There is a lot of time left before 5.5 beta so I would advocate for shipping the correct behavior from the start. If not, are we tracking a follow-up task to finish implementing the expected multi-cursor semantics?

Reply:
Yeah, the last time I spoke to @mjbvz it was decided that for the first version we can keep pasting the same text, but yes, there's definitely time right now before 5.5 beta, so I'll instead implement the correct behavior here.

Reply:
AFAIK multiple copies can only happen with multiple cursors in a single file, but the protocol I suggested seems to imply that it’s possible for the copy sources to be different files. I think this was a mistake in my suggestion 😕

Reply:
I see, so what about something like:
```
export type GetPasteEditsRequestArgs = FileRequestArgs & {
    copies: {
        text: string;
        range?: TextSpan;
    }[];
    copyFromFile?: string;
    pastes: TextSpan[];
};
```
or 
```
export type GetPasteEditsRequestArgs = FileRequestArgs & {
    copiesText: string[];
    copiesFrom?:{ file: string, range: TextSpan[] };
    pastes: TextSpan[];
};
```

Reply:
@mjbvz any input?

Reply:
Proposal:

```ts
export type GetPasteEditsRequestArgs = FileRequestArgs & {
    copiesFrom?:{ file: string, range: TextSpan[] };
    pastedText: string[];
    pastes: TextSpan[];
};
```